File: Exercício Incremental/04/Proprietario.java
import java.util.ArrayList;

public class Proprietario {
  private String nome;
  private String cpf;
  private String identidade;
  private Endereco endereco;
  private ArrayList<Imovel> imoveisDisponiveis;

  public Proprietario(String nome, String cpf, String identidade, String rua, int numero, String cep, String estado, String cidade) {
    this.nome = nome;
    this.cpf = cpf;
    this.identidade = identidade;
    this.endereco = new Endereco(rua, numero, cep, estado, cidade);
    this.imoveisDisponiveis = new ArrayList<Imovel>();
  }

  // Caso não seja informado estado e cidade
  public Proprietario(String nome, String cpf, String identidade, String rua, int numero, String cep) {
    this.nome = nome;
    this.cpf = cpf;
    this.identidade = identidade;
    this.endereco = new Endereco(rua, numero, cep);
    this.imoveisDisponiveis = new ArrayList<Imovel>();
  }

  // GETs
  public String getNome() {
    return this.nome;
  }

  public String getCpf() {
    return this.cpf;
  }

  public String getIdentidade() {
    return this.identidade;
  }

  public String getRua() {
    return this.endereco.getRua();
  }

  public int getNumero() {
    return this.endereco.getNumero();
  }

  public String getCep() {
    return this.endereco.getCep();
  }

  public String getEstado() {
    return this.endereco.getEstado();
  }

  public String getCidade() {
    return this.endereco.getCidade();
  }

  // SETs
  public void setNome(String novoNome) {
    this.nome = novoNome;
  }

  public void setCpf(String novoCpf) {
    this.cpf = novoCpf;
  }

  public void setIdentidade(String novaIdentidade) {
    this.identidade = novaIdentidade;
  }

  public void setRua(String novaRua) {
    this.endereco.setRua(novaRua);
  }

  public void setNumero(int novoNumero) {
    this.endereco.setNumero(novoNumero);
  }

  public void setCep(String novoCep) {
    this.endereco.setCep(novoCep);
  }

  public void setEstado(String novoEstado) {
    this.setEstado(novoEstado);
  }

  public void setCidade(String novaCidade) {
    this.setCidade(novaCidade);
  }

  // Atualizar endereço
  public void atualizaEndereco(String novaRua, int novoNumero, String novoCep, String novoEstado, String novaCidade) {
    this.setRua(novaRua);
    this.setNumero(novoNumero);
    this.setCep(novoCep);
    this.setEstado(novoEstado);
    this.setCidade(novaCidade);
  }

  public void atualizaEndereco(String novaRua, int novoNumero, String novoCep) {
    this.setRua(novaRua);
    this.setNumero(novoNumero);
    this.setCep(novoCep);
  }
  
  // Adicionar imóvel ao proprietário
  public void adicionaImovel(Imovel imovel) {
    this.imoveisDisponiveis.add(imovel);
  }

  // Listar imóveis do proprietário dado seu tipo
  public void listaImoveisPorTipo(String tipo) {
    System.out.println("O proprietário possui os seguintes imóveis do tipo '" + tipo + "':");
    for(Imovel imovel : imoveisDisponiveis) {
      if(imovel.getTipo().equals(tipo)) {
        System.out.println(
          "Iptu: "+imovel.getIptu()+", Utilização: "+imovel.getUtilizacao()+", Cidade: "+imovel.getCidade()
        );
      }
    }
  }
}

File: Trabalho Final/proprietario.py
class Proprietario:
  def __init__(self, nome, cpf, identidade, endereco):
    self.nome = nome
    self.cpf = cpf
    self.identidade = identidade
    self.endereco = endereco
    self.imoveis_disponiveis = []

  def __str__(self):
    return (
      f"Nome: {self.nome}\n" +
      f"CPF: {self.cpf}, RG: {self.identidade}\n" +
      f"Endereço:\n{str(self.endereco)}"
    )

  # Atualizar endereço
  def atualiza_endereco(self, nova_rua, novo_numero, novo_cep, novo_estado=None, nova_cidade=None):
    self.endereco.rua = nova_rua
    self.endereco.numero = novo_numero
    self.endereco.cep = novo_cep
    if (novo_estado is not None):
      self.endereco.set_estado(novo_estado)
    if (nova_cidade is not None):
      self.endereco.set_cidade(nova_cidade)

  # Adicionar imóvel ao proprietário
  def adiciona_imovel(self, imovel):
    self.imoveis_disponiveis.append(imovel)

  # Listar imóveis do proprietário dado seu tipo
  def lista_imoveis_por_tipo(self, tipo):
    print(f"O proprietário possui os seguintes imóveis do tipo '{tipo}': \n")
    for imovel in self.imoveis_disponiveis:
      if (imovel.get_tipo() == tipo):
        print(imovel)
        print()

  # GETs
  def get_nome(self):
    return self.nome

  def get_cpf(self):
    return self.cpf

  def get_identidade(self):
    return self.identidade

  def get_rua(self):
    return self.endereco.get_rua()

  def get_numero(self):
    return self.endereco.get_numero()

  def get_cep(self):
    return self.endereco.get_cep()

  def get_estado(self):
    return self.endereco.get_estado()

  def get_cidade(self):
    return self.endereco.get_cidade()

  # SETs
  def set_nome(self, novo_nome):
    self.nome = novo_nome

  def set_cpf(self, novo_cpf):
    self.cpf = novo_cpf

  def set_identidade(self, nova_identidade):
    self.identidade = nova_identidade

  def set_rua(self, nova_rua):
    self.endereco.set_rua(nova_rua)

  def set_numero(self, novo_numero):
    self.endereco.set_numero(novo_numero)

  def set_cep(self, novo_cep):
    self.endereco.set_cep(novo_cep)

  def set_estado(self, novo_estado):
    self.endereco.set_estado(novo_estado)

  def set_cidade(self, nova_cidade):
    self.endereco.set_cidade(nova_cidade)
